### PR TITLE
Add hook to invoke hardware generators at the end of elaboration.

### PR DESCRIPTION
**Pull request by mikeurbach**
### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- Feature (or new API)

This adds a new API to invoke generators late, within the scope of a Module that has just been elaborated. It is intended as a power user API, and should not normally be needed. However, it does unlock new functionality in some cases where it would otherwise be impossible.

#### Desired Merge Strategy

- Squash: The PR will be squashed and merged (choose this if you have no preference).

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels? (Select the most appropriate one based on the "Type of Improvement")
- [ ] Did you mark the proper milestone (Bug fix: `3.5.x`, `3.6.x`, or `5.x` depending on impact, API modification or big change: `6.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

**Comment by mikeurbach** [comment]
@jackkoenig @mwachs5 @seldridge we discussed an API like this a while back. Here is my take after that discussion, please let me know what you think.

File: core/src/main/scala/chisel3/RawModule.scala [hunk 1]
```diff
@@ -12,13 +12,36 @@ import chisel3.internal.Builder._
...
   */
 @nowarn("msg=class Port") // delete when Port becomes private
 abstract class RawModule extends BaseModule {
+
+  /** Hook to invoke hardware generators after the rest of the Module is constructed.
+    *
+    * This is a power-user API, and should not normally be needed.
+    *
+    * In rare cases, it is necessary to run hardware generators at a late stage, but still within the scope of the
+    * Module. In these situations, atModuleBodyEnd may be used to register such generators. For example:
```
[the remark this answers is not in the document]
Comment (mikeurbach) [reply core/src/main/scala/chisel3/RawModule.scala:29]: Yep, I tried to clarify this in the most recent commit.

File: src/test/scala/chiselTests/RawModuleSpec.scala [hunk 2]
```diff
@@ -68,6 +68,27 @@ class RawModuleSpec extends ChiselFlatSpec with Utils {
     assertTesterPasses({ new RawModuleTester })
   }
 
+  "RawModule" should "support late stage generators with atModuleBodyEnd" in {
```
[the remark this answers is not in the document]
Comment (mikeurbach) [reply src/test/scala/chiselTests/RawModuleSpec.scala]: More or less point-by-point:
* yep- there was already a test for two atModuleBodyEnds, but I added one showing what happens with multiple connects. i also documented what the user can expect from the ordering between atModuleBodyEnd.
* yes, if you try to call DataMirror methods on a module that is not closed, there will be an error, whether you're inside the atModuleBodyEnd or not. atModuleBodyEnd thunks are the last thing to execute before the module is finally closed.
* yep, you can use Definition with these, and the ports exist on the definition. one wrinkle is it doesn't appear possible to make any vals in atModuleBodyEnd @public, so I guess this should be documented.
* yep, you can use DataMirror with these, and the ports exist.

**Comment by mikeurbach** [comment]
I pushed up some documentation and new tests, but the implementation remains the same. Is there anything else I should add in terms documentation or tests? Is this the implementation we want? I believe we discussed a few possible APIs in this vein that @jackkoenig was interested in, so I want to make sure the current implementation is actually one of the ones we want.

**Comment by mikeurbach** [comment]
Also tagging @mwachs5 @azidar @seldridge in case you all have any thoughts on this one.

File: src/test/scala/chiselTests/RawModuleSpec.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (mwachs5) [reply src/test/scala/chiselTests/RawModuleSpec.scala]: >  one wrinkle is it doesn't appear possible to make any vals in atModuleBodyEnd @public, so I guess this should be documented.

If we want to do this then we can make `atModuleBodyEnd` return something which could be marked @public, but fine for now

File: core/src/main/scala/chisel3/RawModule.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (mwachs5) [reply core/src/main/scala/chisel3/RawModule.scala:29]: great thanks

**Review by mwachs5 — APPROVED** [review]
thanks for the tests! lgtm

**Review by jackkoenig — APPROVED** [review]
This API is likely to change a lot in the near future, but it's useful for the time being so let's go ahead and :shipit:

**Pull request merged**